### ISSUE TITLE
Allow for combining for multiple lanes

### ISSUE THREAD

**Issue by jfy133** [nf-core/eager#143]
**Is your feature request related to a problem? Please describe.**
[Despite semi extensive discussion - just realised this isn't described anywhere]

Currently nf-core/nextflow default is to just process together FASTQ files with related Forward/Reverse pairs.

It does not deal with multiple lanes, at least in the way aDNA people normally do, such as a NextSeq where a single library injection is spread across 4 related-lanes of a single flow cell. Or in some cases when a single library is sequenced across multiple lanes of a single or multiple HiSeq flow cells.

We typically want to combine all sets of FASTQs from different lanes together, most importantly prior DeDuplication. This is as we may have PCR duplicates from the same library but across different lanes, and to accurately calculate duplication rate/cluster factor (an important QC metric), we need to account for all duplicates together. 

**TL;DR** Ideally we would like to have a method to indicate how/when to combine FASTQ files from multiple lanes.

**Describe the solution you'd like**
After previous discussion, provide a TSV file which indicates with one column the base name of all the FASTQs to merge, and accordingly which files are associated with which base name.

I personally think everything from FASTQC to samtools filtering can be done on each read pair independently, but combining of BAM files *must* be performed _before_ DeDuplication. This mainly due to metric calculations up until that point can be calculated manually by summing across each FASTQ/BAM, but at this point data generation for the metric needs to have all reads from a library analysed at once.

**Note**, the input channel for Qualimap would have to be changed to account for this (not coming straight from samtools filtering (line 937 of main.nf).

This is something we can discuss though.

**Describe alternatives you've considered**
Some form of flag with a regular expression of how to combine; but due to large heterogeneity in the way files are named at different sequencing centres this was deemed to complicated to account for all cases.

**Additional context**
N/A

**Comment by jfy133** [comment]
I was thinking about this again and I'm thinking again maybe a grouping flag might be a better way to go again. Forcing a user to make a TSV somewhat breaks the flow when running with command line tools. 

While indeed we can't account for the large heterogeneity, we can assume (or rather force) people to have file naming consistency - which you sort of need anyway already with `{R1,R2}` for input files. I don't think it's too harsh to ask people to rename their files if it leads to better organisation.

I would suggest add two flags:

```
--merge_lanes              Turn on combine libraries sequenced across multiple lanes into single BAM for downstream. This occurs prior deduplication.
--merge_lanes_field           Specify the common field of each multi-sequenced library to group together for BAM merging. For example `_L00{1,2,3,4}_`
```

Where, like when you supply the input files as read 1 and read 2, apply that to whatever *consistent* grouping scheme you have already set. You then just `samtools merge` together all the files which have the identical name if you exclude the bash expansion chunk.

Do you think that would be possible? I assume the code (at least in framework form) is half there from the input files going into adapterRemoval.

Other thoughts in this area are:
  * It would be nice to 'ignore' if you have a missing bash expansion file (i.e. if you specify `_L00{1,2,3,4}_` but for one sample only have `_L00{1,2,}_` then it just accepts the `_L00{3,4}_` don't exist.
  * I think it would be safe to say that to _begin_ this would only account for multiple-lanes from the same sequencing run. This would then account at least for default NextSeq output. We can consider elsewise later on (or just require people to rename their files).
  * In some cases you might not be allowed to rename a raw file (e.g. at MPI-SHH) but can 'rename' using a symlink. However AFAIK, nextflow always takes the original file name and not the symlink - so maybe this would be an issue.

Thoughts @apeltzer, @EisenRa, @maxibor, @sc13-bioinf? What are your experiences at your respective institution(s) in terms of FASTQ file naming consistency?

**Comment by sc13-bioinf** [comment]
There is an implementation based on name parsing that I wrote a while ago here:
/projects1/tools/nextflow-workflows/map-to-human.nf 

It also takes care of gathering the stats from adapter removal. The same would probably need to be done for samtools view perhaps. I think that it might need some work so it plays nicely with resume etc. Probably my use of channels and general nextflow could be vastly improved.

**Comment by apeltzer** [comment]
This feature is going to be there simply once we have the #209 metadata sheet information present.

**Comment by jfy133** [comment]
Closing in favour of https://github.com/nf-core/eager/issues/209
